Userland/xargs.cpp

```diff
@@ -0,0 +1,276 @@
+        if (devnull_fd < 0) {
+            perror("open");
+            return 1;
+        }
+    }
+
+    bool fail = false;
+    int lines_used_for_this_command = 0;
+    char* last_line = nullptr;
+    size_t last_line_size = 0;
+    size_t total_command_length = 0;
+
+    for (;;) {
+        line = nullptr;
+        buffer_size = 0;
+
+        ssize_t line_size;
+
+        if (last_line) {
+            line_size = last_line_size;
+            line = last_line;
+            last_line = nullptr;
+        } else {
+            line_size = getdelim(&line, &buffer_size, entry_separator, fp);
+        }
+
+        if (line_size < 0) {
+            if (errno != 0) {
+                perror("getdelim");
+                fail = true;
+            }
+            break;
+        }
+
+        last_line = line;
+        last_line_size = line_size;
+
+        if (line_size == 0) {
+            free(line);
+            last_line = nullptr;
+            continue;
+        }
+
+        ++lines_used_for_this_command;
+
+        if (line_size > 0) {
+            while (line[line_size - 1] == '\n')
+                --line_size;
+
+            if (!line_size) {
+                last_line = nullptr;
+                continue;
+            }
+
+            StringView line_view { line, (size_t)line_size };
+
+            dbg() << "Getline '" << line_view << "'";
+
+            size_t index = 0;
+            if (lines_used_for_this_command == 1) {
+                child_argv.ensure_capacity(argument_parts.size());
+                for (auto& parts : argument_parts) {
+                    builder.clear();
+                    builder.join(line_view, parts);
+                    auto string = builder.build();
+                    total_command_length += string.length();
+                    child_argv.unchecked_append(strdup(string.characters()));
+                    ++index;
+                }
+            } else {
+                if ((max_lines_for_one_command > 0 && lines_used_for_this_command > max_lines_for_one_command) || total_command_length + line_size + 1 >= max_bytes) {
+                    run_command(child_argv, verbose, is_stdin, devnull_fd);
+                    lines_used_for_this_command = 0;
+                    total_command_length = 0;
+                    continue;
+                } else {
+                    child_argv.append(strndup(line, line_size));
+                }
+            }
+
+            free(line);
+            last_line = nullptr;
+        }
+    }
+
+    if (!child_argv.is_empty())
+        run_command(child_argv, verbose, is_stdin, devnull_fd);
+
+    if (!is_stdin)
+        fclose(fp);
+
+    return fail ? 1 : 0;
+}
+
+int run_command(Vector<char*>& child_argv, bool verbose, bool is_stdin, bool devnull_fd)
+{
+    child_argv.append(0);
+
+    if (verbose) {
+        fprintf(stderr, "xargs: ");
+        for (auto* ptr : child_argv) {
+            if (!ptr)
+                break;
+            fprintf(stderr, "%s ", ptr);
```

Comment:
Super nitpick: not cool to output a space after the last arg; perhaps use `join()` or something.

Reply:
\*cough\* it's not even visible